sky/packages/sky/lib/src/fn3/framework.dart

```diff
@@ -686,21 +684,27 @@ abstract class BuildableElement<T extends Widget> extends Element<T> {
 
   static BuildScheduler scheduleBuildFor;
 
-  static int _debugStateLockLevel = 0;
-  static bool get _debugStateLocked => _debugStateLockLevel > 0;
-
-  /// Calls the callback argument synchronously, but in a context where calls to
-  /// State.setState() will fail. Use this when it is possible that you will
-  /// trigger code in components but want to make sure that there is no
-  /// possibility that any components will be marked dirty, for example because
-  /// you are in the middle of layout and you are not going to be flushing the
-  /// build queue (since that could mutate the layout tree).
-  static void lockState(void callback()) {
-    _debugStateLockLevel += 1;
+  static int _debugBuildScopeDepth = 0;
+  static bool get _debugInBuildScope => _debugBuildScopeDepth > 0;
+
+  /// Establishes a scope in which component build functions can run.
+  ///
+  /// Inside a build scope, component build functions are allowed to run, but
+  /// State.setState() is forbidden. This mechanism prevents build functions
+  /// from transitively requiring other build functions to run, potentially
+  /// causing infinite loops.
+  ///
+  /// After unwinding the last build scope on the stack, the framework verifies
+  /// that each global key is used at most once and notifies listeners about
+  /// changes to global keys.
+  static void buildScope(void callback()) {
+    _debugBuildScopeDepth += 1;
     try {
       callback();
+      if (_debugBuildScopeDepth == 1)
+        GlobalKey._checkForDuplicatesAndNotifyListeners();
```

Comment:
come to think of it, this still won't handle cases where you move a node from a MixedViewport list to something that builds before that list. That update is done in two steps: one buildScope() for the "main" pass, and one for the layout() of the MixedViewport (which, btw, means that if something deeper in the MixedViewport was dirty, we'll update it before the list, which might be a wasted update... oh well), and one during layout. The check in the middle will see a duplicate GlobalKeys, even though by the end of layout it won't be duplicate any more.